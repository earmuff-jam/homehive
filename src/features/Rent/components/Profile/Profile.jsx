import React, { useEffect, useState } from "react";

import { Controller, useForm } from "react-hook-form";

import dayjs from "dayjs";

import { InfoRounded } from "@mui/icons-material";
import {
  Avatar,
  Box,
  Card,
  Chip,
  Grid,
  Skeleton,
  Stack,
  Typography,
} from "@mui/material";
import AButton from "common/AButton";
import CustomSnackbar from "common/CustomSnackbar/CustomSnackbar";
import RowHeader from "common/RowHeader/RowHeader";
import TextFieldWithLabel from "common/TextFieldWithLabel";
import relativeTime from "dayjs/plugin/relativeTime";
import {
  useGetUserDataByIdQuery,
  useUpdateUserByUidMutation,
} from "features/Api/firebaseUserApi";
import { fetchLoggedInUser } from "features/Rent/utils/utils";

dayjs.extend(relativeTime);

export const Profile = () => {
  const user = fetchLoggedInUser();

  const { data: userData, isLoading: isUserDataLoading } =
    useGetUserDataByIdQuery(user?.uid, {
      skip: !user?.uid,
    });

  const [
    updateUser,
    { isSuccess: isUpdateUserSuccess, isLoading: isUpdateUserLoading },
  ] = useUpdateUserByUidMutation();

  const [showSnackbar, setShowSnackbar] = useState(false);

  const {
    control,
    handleSubmit,
    reset,
    formState: { errors, isValid },
  } = useForm({
    mode: "onChange",
    defaultValues: {
      first_name: "",
      last_name: "",
      phone: "",
      street_address: "",
      city: "",
      state: "",
      zipcode: "",
    },
  });

  const onSubmit = (formData) => {
    updateUser({
      uid: user?.uid,
      newData: {
        ...formData,
        uid: user?.uid,
        googleAccountLinkedAt: userData?.googleAccountLinkedAt,
        googleDisplayName: userData?.googleDisplayName,
        googleEmailAddress: userData?.googleEmailAddress,
        googleLastLoginAt: userData?.googleLastLoginAt,
        googlePhotoURL: userData?.googlePhotoURL,
        updatedOn: dayjs().toISOString(),
        updatedBy: user?.uid,
      },
    });
    setShowSnackbar(true);
  };

  useEffect(() => {
    if (isUpdateUserSuccess) setShowSnackbar(false);
  }, [isUpdateUserSuccess]);

  useEffect(() => {
    if (userData) {
      reset({
        first_name: userData?.first_name || "",
        last_name: userData?.last_name || "",
        phone: userData?.phone || "",
        street_address: userData?.street_address || "",
        city: userData?.city || "",
        state: userData?.state || "",
        zipcode: userData?.zipcode || "",
      });
    }
  }, [isUserDataLoading, reset]);

  if (isUserDataLoading) return <Skeleton height="10rem" />;

  return (
    <Grid container spacing={3}>
      <Grid item xs={12} md={4}>
        <Card elevation={0} sx={{ padding: 3, textAlign: "center" }}>
          <Avatar
            sx={{
              width: 120,
              height: 120,
              mx: "auto",
              fontSize: "2.5rem",
              bgcolor: "primary.main",
            }}
            src={userData?.googlePhotoURL || ""}
          />

          <Typography variant="h6" fontWeight={600} color="textSecondary">
            {userData.googleDisplayName}
          </Typography>
          <Typography variant="h6" fontWeight={600} color="textSecondary">
            {userData.googleEmailAddress}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            {user?.role}
          </Typography>
          <Chip label="Verified" color="primary" size="small" sx={{ mt: 1 }} />
        </Card>
      </Grid>

      <Grid item xs={12} md={8}>
        <form onSubmit={handleSubmit(onSubmit)}>
          <Card
            elevation={0}
            sx={{
              display: "flex",
              flexDirection: "column",
              gap: "1rem",
              padding: 2,
            }}
          >
            <RowHeader
              title="Personal Information"
              sxProps={{
                textAlign: "left",
                fontWeight: "bold",
                color: "text.secondary",
              }}
            />

            {/* First and Last Name */}
            <Stack direction="row" spacing={2}>
              <Controller
                name="first_name"
                control={control}
                rules={{
                  required: "First name is required",
                  validate: (value) =>
                    value.trim().length > 3 ||
                    "First name must be more than 3 characters",
                  maxLength: {
                    value: 150,
                    message: "First name should be less than 150 characters",
                  },
                }}
                render={({ field }) => (
                  <TextFieldWithLabel
                    {...field}
                    label="First Name *"
                    id="first_name"
                    name="first_name"
                    placeholder="Enter your first name"
                    error={!!errors.first_name}
                    errorMsg={errors.first_name?.message}
                    fullWidth
                  />
                )}
              />

              <Controller
                name="last_name"
                control={control}
                rules={{
                  required: "Last name is required",
                  validate: (value) =>
                    value.trim().length > 0 || "Last name is required",
                  maxLength: {
                    value: 150,
                    message: "Last name should be less than 150 characters",
                  },
                }}
                render={({ field }) => (
                  <TextFieldWithLabel
                    {...field}
                    label="Last Name *"
                    id="last_name"
                    name="last_name"
                    placeholder="Enter your Last Name"
                    error={!!errors.last_name}
                    errorMsg={errors.last_name?.message}
                    fullWidth
                  />
                )}
              />
            </Stack>

            {/* Email and Phone Number */}
            <Stack direction="row" spacing={2}>
              <TextFieldWithLabel
                label="Email address *"
                id="email"
                name="email"
                placeholder="Email Address"
                value={userData?.googleEmailAddress || ""}
                isDisabled
                errorMsg=""
                labelIcon={<InfoRounded fontSize="small" color="secondary" />}
                labelIconHelper="Editing an email address is prevented by default."
              />

              <Controller
                name="phone"
                control={control}
                rules={{
                  required: "Phone number is required",
                  pattern: {
                    value: /^\d{10}$/,
                    message: "Phone number must be a valid 10-digit number",
                  },
                }}
                render={({ field }) => (
                  <TextFieldWithLabel
                    {...field}
                    label="Phone Number *"
                    id="phone"
                    name="phone"
                    placeholder="Enter your phone number"
                    error={!!errors.phone}
                    errorMsg={errors.phone?.message}
                  />
                )}
              />
            </Stack>

            {/* Street Address */}
            <Controller
              name="street_address"
              control={control}
              rules={{
                required: "Street address is required",
                maxLength: {
                  value: 150,
                  message: "Street address should be less than 150 characters",
                },
              }}
              render={({ field }) => (
                <TextFieldWithLabel
                  {...field}
                  label="Street Address *"
                  id="street_address"
                  name="street_address"
                  placeholder="Enter your primary street address"
                  error={!!errors.street_address}
                  errorMsg={errors.street_address?.message}
                />
              )}
            />

            {/* City, State, Zip Code */}
            <Stack direction="row" spacing={2}>
              <Controller
                name="city"
                control={control}
                rules={{
                  required: "City is required",
                  maxLength: {
                    value: 150,
                    message: "City should be less than 150 characters",
                  },
                }}
                render={({ field }) => (
                  <TextFieldWithLabel
                    {...field}
                    label="City *"
                    id="city"
                    name="city"
                    placeholder="City"
                    error={!!errors.city}
                    errorMsg={errors.city?.message}
                  />
                )}
              />
              <Controller
                name="state"
                control={control}
                rules={{
                  required: "State is required",
                  minLength: {
                    value: 2,
                    message: "State is required in the form of XX. Eg, AZ",
                  },
                  maxLength: {
                    value: 2,
                    message: "State is required in the form of XX. Eg, AZ",
                  },
                }}
                render={({ field }) => (
                  <TextFieldWithLabel
                    {...field}
                    label="State *"
                    id="state"
                    name="state"
                    placeholder="State"
                    error={!!errors.state}
                    errorMsg={errors.state?.message}
                  />
                )}
              />
              <Controller
                name="zipcode"
                control={control}
                rules={{
                  required: "Zip Code is required",
                  pattern: {
                    value: /^\d{5}$/,
                    message: "Zip Code should be exactly 5 digits",
                  },
                }}
                render={({ field }) => (
                  <TextFieldWithLabel
                    {...field}
                    label="Zip Code *"
                    id="zipcode"
                    name="zipcode"
                    placeholder="Zip Code"
                    error={!!errors.zipcode}
                    errorMsg={errors.zipcode?.message}
                  />
                )}
              />
            </Stack>

            <Box>
              <AButton
                label="Save"
                variant="contained"
                type="submit"
                disabled={!isValid}
                loading={isUpdateUserLoading}
              />
            </Box>
            <Typography variant="caption" sx={{ fontStyle: "italic" }}>
              Last login around&nbsp;
              {dayjs(userData?.googleLastLoginAt).fromNow() ||
                dayjs().fromNow()}
            </Typography>
          </Card>
        </form>
      </Grid>
      <CustomSnackbar
        showSnackbar={showSnackbar}
        setShowSnackbar={setShowSnackbar}
        title="Changes saved."
      />
    </Grid>
  );
};
