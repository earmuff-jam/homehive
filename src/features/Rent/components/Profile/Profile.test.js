import React from "react";

import { Provider } from "react-redux";

import { Profile } from "./Profile";
import { configureStore } from "@reduxjs/toolkit";
import { render, screen } from "@testing-library/react";

jest.mock("features/Api/firebaseUserApi", () => ({
  useGetUserDataByIdQuery: jest.fn(() => ({
    data: {
      uid: "123",
      googleDisplayName: "John Doe",
      googleEmailAddress: "john@example.com",
      googlePhotoURL: "",
      googleLastLoginAt: new Date().toISOString(),
    },
    isLoading: false,
  })),
  useUpdateUserByUidMutation: jest.fn(() => [jest.fn(), {}]),
}));

jest.mock("features/Rent/utils/utils", () => ({
  fetchLoggedInUser: jest.fn(() => ({
    uid: "123",
    role: "Owner",
  })),
}));

jest.mock("common/TextFieldWithLabel", () => (props) => (
  <input data-testid={props.id} {...props} />
));
jest.mock("common/AButton", () => (props) => (
  <button data-testid="save-button">{props.label}</button>
));
jest.mock("common/CustomSnackbar/CustomSnackbar", () => () => (
  <div data-testid="snackbar" />
));
jest.mock("common/RowHeader/RowHeader", () => (props) => (
  <div data-testid="row-header">{props.title}</div>
));

const store = configureStore({
  reducer: (state = {}) => state,
});

describe("<Profile />", () => {
  it("should render and test snapshots", () => {
    const { asFragment } = render(<Profile />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render without crashing and shows main sections", () => {
    render(
      <Provider store={store}>
        <Profile />
      </Provider>,
    );

    expect(screen.getByText("John Doe")).toBeInTheDocument();
    expect(screen.getByText("john@example.com")).toBeInTheDocument();
    expect(screen.getByTestId("row-header")).toBeInTheDocument();
    expect(screen.getByTestId("save-button")).toBeInTheDocument();
  });
});
