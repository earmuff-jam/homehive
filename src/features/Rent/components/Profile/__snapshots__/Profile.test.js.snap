// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> should render and test snapshots 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-1qw2a80-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lj5egr-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-e9t90p-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: none;"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-ofhj4p-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-3zp90o-MuiTypography-root"
        >
          John Doe
        </h6>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-3zp90o-MuiTypography-root"
        >
          john@example.com
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          Owner
        </p>
        <div
          class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-filledPrimary css-1nt9eff-MuiChip-root"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
          >
            Verified
          </span>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-9drkqd-MuiGrid-root"
    >
      <form>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-o3b9az-MuiPaper-root-MuiCard-root"
          style="--Paper-shadow: none;"
        >
          <div
            data-testid="row-header"
          >
            Personal Information
          </div>
          <div
            class="MuiStack-root css-pl8gqh-MuiStack-root"
          >
            <input
              data-testid="first_name"
              id="first_name"
              label="First Name *"
              name="first_name"
              placeholder="Enter your first name"
              value=""
            />
            <input
              data-testid="last_name"
              id="last_name"
              label="Last Name *"
              name="last_name"
              placeholder="Enter your Last Name"
              value=""
            />
          </div>
          <div
            class="MuiStack-root css-pl8gqh-MuiStack-root"
          >
            <input
              data-testid="email"
              errormsg=""
              id="email"
              label="Email address *"
              labelicon="[object Object]"
              labeliconhelper="Editing an email address is prevented by default."
              name="email"
              placeholder="Email Address"
              value="john@example.com"
            />
            <input
              data-testid="phone"
              id="phone"
              label="Phone Number *"
              name="phone"
              placeholder="Enter your phone number"
              value=""
            />
          </div>
          <input
            data-testid="street_address"
            id="street_address"
            label="Street Address *"
            name="street_address"
            placeholder="Enter your primary street address"
            value=""
          />
          <div
            class="MuiStack-root css-pl8gqh-MuiStack-root"
          >
            <input
              data-testid="city"
              id="city"
              label="City *"
              name="city"
              placeholder="City"
              value=""
            />
            <input
              data-testid="state"
              id="state"
              label="State *"
              name="state"
              placeholder="State"
              value=""
            />
            <input
              data-testid="zipcode"
              id="zipcode"
              label="Zip Code *"
              name="zipcode"
              placeholder="Zip Code"
              value=""
            />
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <button
              data-testid="save-button"
            >
              Save
            </button>
          </div>
          <span
            class="MuiTypography-root MuiTypography-caption css-c289yo-MuiTypography-root"
          >
            Last login aroundÂ a few seconds ago
          </span>
        </div>
      </form>
    </div>
    <div
      data-testid="snackbar"
    />
  </div>
</DocumentFragment>
`;
